pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java

```diff
@@ -0,0 +1,121 @@
+package engineering.everest.lhotse.registrations.domain;
+
+import engineering.everest.lhotse.organizations.domain.commands.CreateRegisteredOrganizationCommand;
+import engineering.everest.lhotse.registrations.domain.commands.CancelConfirmedRegistrationUserEmailAlreadyInUseCommand;
+import engineering.everest.lhotse.registrations.domain.commands.CompleteOrganizationRegistrationCommand;
+import engineering.everest.lhotse.registrations.domain.commands.RecordSentOrganizationRegistrationEmailConfirmationCommand;
+import engineering.everest.lhotse.registrations.domain.events.OrganizationRegistrationCompletedEvent;
+import engineering.everest.lhotse.registrations.domain.events.OrganizationRegistrationConfirmedAfterUserWithEmailCreatedEvent;
+import engineering.everest.lhotse.registrations.domain.events.OrganizationRegistrationConfirmedEvent;
+import engineering.everest.lhotse.registrations.domain.events.OrganizationRegistrationReceivedEvent;
+import engineering.everest.lhotse.users.domain.commands.CreateUserForNewlyRegisteredOrganizationCommand;
+import engineering.everest.lhotse.users.domain.commands.PromoteUserToOrganizationAdminCommand;
+import engineering.everest.lhotse.users.domain.events.UserCreatedForNewlyRegisteredOrganizationEvent;
+import engineering.everest.lhotse.users.services.UsersReadService;
+import lombok.NoArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.axonframework.commandhandling.gateway.CommandGateway;
+import org.axonframework.modelling.saga.EndSaga;
+import org.axonframework.modelling.saga.SagaEventHandler;
+import org.axonframework.modelling.saga.StartSaga;
+import org.axonframework.serialization.Revision;
+import org.axonframework.spring.stereotype.Saga;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+import java.util.UUID;
+
+@Saga
+@Revision("0")
+@Log4j2
+@NoArgsConstructor
+public class OrganizationRegistrationSaga {
```

Comment:
I still don't understand why Saga cannot be cached. Axon doc does have a brief [section](https://docs.axoniq.io/reference-guide/configuring-infrastructure-components/event-processing/saga-infrastructure#caching) talking about it. Could you please help me understand why it cannot be used in our case?

Reply:
It's a limitation of us relying on Hazelcast to assign aggregate ownership based on their IDs. In this saga, all commands dispatched to the organization aggregate are handled by the same node but the user aggregate commands are likely to be handled by different nodes.

When this happens, the second node loads the state of the saga from the database. The first node, however, doesn't seem to have any way of detecting that its version of the saga is out of date (from what I saw).

This happens to be safe for this saga as the saga state is not being changed as it's just reacting to events and dispatching commands but it seems like to cache it would be setting a trap for people.

Reply:
Thanks for the explanation. This seems to be a major issue. Because Saga is meant to be cross aggregate and with Lhotse's setup, cross aggregate means multiple nodes. This basically makes Saga pretty useless as it cannot be stateful?

Is there any possible solution around this? Somehow route this event to the already loaded Saga instance in a way similar to distributed command is handled?

Reply:
It limits the ability to cache but it should still possible to hydrate the state from the DB every time an event is handled (the default)

Reply:
ok so its not a deal breaker. Phew ...